### PR TITLE
Add support for `ViewComponentTagHelpers`.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Hardcoded `ViewComponent` discovery.
- Hardcoded `ViewComponentTagHelperDescriptor` creation.
- Added test to validate that ViewComponents are discovered and transitioned into TagHelpers properly.
- Avoided adding a reference to MVC to prevent circular references. This resulted in custom marker attributes to represent `ViewComponent`s. Also made a lot of use of `ViewComponent` conventions (ending in "ViewComponent").

#932

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs [hunk 1]
```diff
@@ -19,46 +21,82 @@ public DefaultTagHelperResolver(bool designTime)
         public override IReadOnlyList<TagHelperDescriptor> GetTagHelpers(Compilation compilation)
         {
             var results = new List<TagHelperDescriptor>();
+            var errors = new ErrorSink();
```
Comment (NTaylorMullen) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs]: We don't flow this error sink out today. It'd probably be valuable to do that to inform tooling on bits that go wrong.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs [hunk 2]
```diff
@@ -19,46 +21,82 @@ public DefaultTagHelperResolver(bool designTime)
...
+
+            VisitCompilation(visitor, compilation);
+
+            var factory = new ViewComponentTagHelperDescriptorFactory(compilation);
+
+            foreach (var type in types)
+            {
+                try
+                {
+                    var descriptor = factory.CreateDescriptor(type);
```
Comment (NTaylorMullen) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs]: These bits will be replaced by MVC invocations. Didn't feel that it was worth the time to convert each exception clause into an `errorSink.OnError` call/flow the ErrorSink throughout the factory.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs]: Yep. We need to do a refactor on `THD` or least evaluate the design. Let's plan on doing this today.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs [hunk 3]
```diff
@@ -19,46 +21,82 @@ public DefaultTagHelperResolver(bool designTime)
...
             {
                 var descriptors = factory.CreateDescriptors(type, errors);
                 results.AddRange(descriptors);
             }
+        }
 
-            return results;
+        private static void VisitViewComponents(Compilation compilation, List<TagHelperDescriptor> results, ErrorSink errors)
+        {
+            var @interface = compilation.GetTypeByMetadataName(ViewComponentTypes.ViewComponentAttribute);
```
Comment (rynowak) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs]: should we no-op if this type isn't defined?

Comment (rynowak) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs]: Have you given any thought to how we'll determine if the user is on `MVC >= 1.1`

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs [hunk 4]
```diff
@@ -88,5 +126,68 @@ internal bool IsTagHelper(INamedTypeSymbol symbol)
...
+            internal bool IsViewComponent(INamedTypeSymbol symbol)
+            {
+                if (symbol.DeclaredAccessibility != Accessibility.Public ||
+                    symbol.IsAbstract ||
+                    symbol.IsGenericType)
+                {
+                    return false;
+                }
+
+                return symbol.Name.EndsWith(ViewComponentTypes.ViewComponentSuffix) ||
```
Comment (rynowak) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs]: There's also `NonViewComponent` attribute to consider, which makes the entire hierarchy not a VC.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ViewComponentTagHelperDescriptorFactory.cs [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+                        componentType.ToDisplayString(FullNameTypeDisplayFormat),
+                        nameof(Task)));
+                }
+            }
+            else if (returnType != null)
+            {
+                // Will invoke synchronously. Method must not return void, Task or Task<T>.
+                if (returnType.SpecialType == SpecialType.System_Void)
+                {
+                    throw new InvalidOperationException(ViewComponentResources.FormatViewComponent_SyncMethod_ShouldReturnValue(
```
Comment (rynowak) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/ViewComponentTagHelperDescriptorFactory.cs]: Are these the errors the VC would report at runtime when called?

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs (hunk 3 above)
Comment (NTaylorMullen) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs]: > should we no-op if this type isn't defined?

If there's no attribute the classes could be defined in a class library using the ViewComponent suffix couldn't they?



> Have you given any thought to how we'll determine if the user is on MVC >= 1.1

Great question. Hmm, I could look for public types added in 1.1 to check for its existence. What do you think?

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ViewComponentTagHelperDescriptorFactory.cs (hunk 5 above)
Comment (NTaylorMullen) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/ViewComponentTagHelperDescriptorFactory.cs]: Yes

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs (hunk 4 above)
Comment (NTaylorMullen) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs]: Oh shoot, good catch. I missed that bit.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs (hunk 3 above)
Comment (NTaylorMullen) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs]: Hmm, depending on the approach for detecting 1.1 MVC maybe nooping would be the right approach

**Pull request merged**